app/src/main/java/com/internship/move/ui/authentication/AuthenticationViewModel.kt

```diff
@@ -1,23 +1,76 @@
 package com.internship.move.ui.authentication
 
+import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
-import com.internship.move.repository.Repository
+import com.internship.move.network.model.AddLicenseResponse
+import com.internship.move.network.model.ErrorResponse
+import com.internship.move.network.model.LoginResponse
+import com.internship.move.network.model.RegisterResponse
+import com.internship.move.repository.UserRepository
+import com.internship.move.utils.Constants.LOADING_DELAY
+import com.internship.move.utils.InternalStorageManager
+import com.squareup.moshi.Moshi
+import kotlinx.coroutines.delay
+import org.koin.java.KoinJavaComponent.inject
+import retrofit2.HttpException
 
 class AuthenticationViewModel(
-    private val repo: Repository
+    private val repo: UserRepository,
+    private val internalStorageManager: InternalStorageManager
 ) : ViewModel() {
 
-    fun register(email: String, username: String, password: String) {
-        //api call
-        repo.setIsLoggedIn(true)
-    }
+    private val moshi: Moshi by inject(Moshi::class.java)
+    val errorMessage: MutableLiveData<String> = MutableLiveData()
+    val isLoading: MutableLiveData<Boolean> = MutableLiveData(false)
 
-    fun login(email: String, password: String) {
-        // api call
-        repo.setIsLoggedIn(true)
-    }
+    suspend fun register(name: String, email: String, password: String): RegisterResponse? =
```

Comment:
Instead of returning values, use LiveData and its observer pattern to get the updates in your Fragment.
This way you won't have to return `null` in cath (or anything). Also, by doing this, you can remove the `suspend` operator and launch a coroutine using the `viewModelScope.launch { }` method here in the viewModel, instead of launching it form the fragment. You want api calls to be boun do the viewModel's "lifecycle"

Reply:
Also, you should stop the loading if the api call succeeds as well :P

Reply:
Apply these changes wherever necessary in the other VMs/Fragments too